src/api/che-api.ts

```diff
@@ -0,0 +1,340 @@
+/*********************************************************************
+ * Copyright (c) 2019-2020 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import { che as chetypes } from '@eclipse-che/api'
+import axios, { AxiosInstance } from 'axios'
+import { cli } from 'cli-ux'
+import * as https from 'https'
+
+/**
+ * Singleton responsible for calls to Che API.
+ */
+let instance: CheApiClient | undefined
+export class CheApiClient {
+  public defaultCheResponseTimeoutMs = 3000
+  public readonly cheApiUrl: string
+
+  private readonly axios: AxiosInstance
+
+  private constructor(cheApiUrl: string) {
+    this.cheApiUrl = cheApiUrl
+
+    // Make axios ignore untrusted certificate error for self-signed certificate case.
+    const httpsAgent = new https.Agent({ rejectUnauthorized: false })
+
+    this.axios = axios.create({
+      httpsAgent
+    })
+  }
+
+  public static getInstance(cheApiUrl: string): CheApiClient {
+    cheApiUrl = this.normalizeCheApiUrl(cheApiUrl)!
+    if (!instance || instance.cheApiUrl !== cheApiUrl) {
+      instance = new CheApiClient(cheApiUrl)
+    }
+    return instance
+  }
+
+  private static normalizeCheApiUrl(url: string | undefined) {
+    if (url) {
+      if (!url.includes('://')) {
+        url = 'https://' + url
+      }
+      const u = new URL(url)
+      url = 'https://' + u.host + u.pathname
+      if (url.endsWith('/')) {
+        url = url.slice(0, -1)
+      }
+      return url
+    }
+  }
+
+  /**
+   * Checks whether provided url really points to Che server API.
+   * Trows an exception if not.
+   */
+  async ensureCheApiUrlCorrect(responseTimeoutMs = this.defaultCheResponseTimeoutMs): Promise<void> {
+    try {
+      const response = await this.axios.get(`${this.cheApiUrl}/system/state`, { timeout: responseTimeoutMs })
+      if (response.data && response.data.status) {
+        return
+      }
+    } catch {
+      throw new Error(`E_CHE_API_URL_NO_RESPONSE - Failed to connect to "${this.cheApiUrl}". Is it the right url?`)
+    }
+    throw new Error(`E_CHE_API_WRONG_URL - Provided url "${this.cheApiUrl}" is not Che API url`)
+  }
+
+  async isCheServerReady(responseTimeoutMs = this.defaultCheResponseTimeoutMs): Promise<boolean> {
+    const id = await this.axios.interceptors.response.use(response => response, async (error: any) => {
+      if (error.config && error.response && (error.response.status === 404 || error.response.status === 503)) {
+        return this.axios.request(error.config)
```

Comment:
I guess this code made requests while Che will be ready. Do we need timeout for this interceptor recursion?

Reply:
Yes, good catch, thanks.